sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/FrameworkScheduler.java

```diff
@@ -0,0 +1,269 @@
+     * started, because when tasks launch they typically require access to ArtifactResource for config templates.
+     */
+    private final AtomicBoolean readyToAcceptOffers = new AtomicBoolean(false);
+
+    private final Set<String> frameworkRolesWhitelist;
+    private final FrameworkStore frameworkStore;
+    private final AbstractScheduler abstractScheduler;
+    private final OfferProcessor offerProcessor;
+
+    private final StateStore stateStore;
+    private TaskCleaner taskCleaner;
+    private boolean multithreaded = true;
+
+    public FrameworkScheduler(
+            Set<String> frameworkRolesWhitelist,
+            SchedulerConfig schedulerConfig,
+            StateStore stateStore,
+            FrameworkStore frameworkStore,
+            AbstractScheduler abstractScheduler) {
+        this(
+                frameworkRolesWhitelist,
+                frameworkStore,
+                abstractScheduler,
+                new OfferProcessor(abstractScheduler, stateStore),
+                stateStore);
+    }
+
+    @VisibleForTesting
+    FrameworkScheduler(
+            Set<String> frameworkRolesWhitelist,
+            FrameworkStore frameworkStore,
+            AbstractScheduler abstractScheduler,
+            OfferProcessor offerProcessor,
+            StateStore stateStore) {
+        this.frameworkRolesWhitelist = frameworkRolesWhitelist;
+        this.frameworkStore = frameworkStore;
+        this.abstractScheduler = abstractScheduler;
+        this.offerProcessor = offerProcessor;
+        this.stateStore = stateStore;
+    }
+
+    /**
+     * Notifies this instance that the API server has been initialized. All offers are declined until this is called.
+     *
+     * @return {@code this}
+     */
+    public FrameworkScheduler setReadyToAcceptOffers() {
+        readyToAcceptOffers.set(true);
+        return this;
+    }
+
+    /**
+     * Disables multithreading for tests. For this to take effect, it must be invoked before the framework has
+     * registered.
+     *
+     * @return {@code this}
+     */
+    @VisibleForTesting
+    public FrameworkScheduler disableThreading() {
+        offerProcessor.disableThreading();
+        this.multithreaded = false;
+        return this;
+    }
+
+    @Override
+    public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, Protos.MasterInfo masterInfo) {
+        if (isRegisterStarted.getAndSet(true)) {
+            // This may occur as the result of a master election.
+            LOGGER.info("Already registered, calling reregistered()");
+            reregistered(driver, masterInfo);
+            return;
+        }
+
+        Driver.setDriver(driver);
+        abstractScheduler.registeredWithMesos();
+
+        LOGGER.info("Registered framework with frameworkId: {}", frameworkId.getValue());
+        this.taskCleaner = new TaskCleaner(stateStore, multithreaded);
```

Comment:
Does this need to be initialized post registration?  StateStore is available on construction.

Reply:
`multithreaded` may be set false in tests

Reply:
Slightly unfortunate.  Can threading be enabled/disabled on construction or do tests require doing it post construction?  Is it a convenience thing?

Reply:
It's mainly for convenience -- tests may disable threading at the AbstractScheduler level and then it cascades into here. For what it's worth this will end up going away when FrameworkScheduler is finished being built out, this part is in an interim state with this PR.